DataDog/datadog-operator
# [CECO-884] Add sidecar injection basic config in operator

### What does this PR do?

Add Agent sidecar injection feature with basic config in fargate environment.
Also refactored admissionController feature test codes.

### Motivation

CECO-884

### Additional Notes

Extending agent sidecar injection from [helm](https://github.com/DataDog/helm-charts/blob/97d9522f38163b4697b714927be0a9a292bd5085/charts/datadog/values.yaml#L1109) to operator. 

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: v7.53.0
* Cluster Agent: v7.53.0

### Describe your test plan

Test label overrides with < 1.7.0. (e.g. 1.6.0)

- Create DDA with below mandatory setup :
    - `features.admissionController.agentSidecarInjection.enabled` to true.
    - `features.admissionController.agentSidecarInjection` to “fargate” as minimal setup.

```
spec: 
  features:
    admissionController:
      agentSidecarInjection:
        enabled: true
        provider: fargate
```

- Once you deploy it, Check the environment variable is set below.
- below value is default variables with minimal setup.
- image name and image tag will follow the default agent image name and tag.
- In operator 1.6.0, DDA 7.53.0 is default version tag.

```
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED:                true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CLUSTER_AGENT_ENABLED:  true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER:               fargate
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME:             agent
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG:              7.53.0
```

- Check the sidecar agent has correctly attached with applications.

```
  datadog-agent-injected:
    Container ID:   containerd://b30f4e833aa7bfa51a7c183ffed54b3a14f75f2632bd1aad6353fd207c5c17ea
    Image:          gcr.io/datadoghq/agent:7.53.0
```

- Test with extra features (e.g. registry, imageName, imageTag) and See the values are set properly.

```
  features:
    admissionController:
      agentSidecarInjection:
        enabled: true
        provider: fargate
        registry: public.ecr.aws/datadog
        imageName: agent
        imageTag: 7.52.0
```

- Check the result

```
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED:                true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CLUSTER_AGENT_ENABLED:  true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER:               fargate
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CONTAINER_REGISTRY:     public.ecr.aws/datadog
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME:             agent
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG:              7.52.0
```

- Check the agent images is pulling the right path, name and version.

```
  datadog-agent-injected:
    Container ID:   
    Image:          public.ecr.aws/datadog/agent:7.52.0
```
- Lastly, Test with nodeAgent override image name and tag and it will override default image name and tag.
- and Also see If there is value of them in admissionController, It will take precedence.

When only set `override`
```
  override:
    nodeAgent:
      image:
        name: NameFromOverride
        tag: TagFromOverride
```
Result : See the configuration from override.
```
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED:                true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CLUSTER_AGENT_ENABLED:  true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER:               fargate
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CONTAINER_REGISTRY:     public.ecr.aws/datadog
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME:             NameFromOverride
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG:              TagFromOverride
```

When both set `override` and `feature`
```
  override:
    nodeAgent:
      image:
        name: NameFromOverride
        tag: TagFromOverride
...
  features:
    admissionController:
      agentSidecarInjection:
        enabled: true
        provider: fargate
        registry: public.ecr.aws/datadog
        imageName: NameFromFeature
        imageTag: TagFromFeature
```
Result : See the configuration from feature will take precedence.
```
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED:                true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CLUSTER_AGENT_ENABLED:  true
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER:               fargate
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_CONTAINER_REGISTRY:     public.ecr.aws/datadog
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_NAME:             NameFromFeature
      DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_IMAGE_TAG:              TagFromFeature
```

### Checklist
